pyupdi.py

```diff
@@ -48,53 +51,88 @@
+    parser.add_argument("-fs", "--fuses", action="append", nargs="*",
+                        help="Fuse to set (syntax: fuse_nr:0xvalue)")
+    parser.add_argument("-v", "--verbose", action="store_true",
+                        help="Set verbose mode")
+
+    args = parser.parse_args(sys.argv[1:])
+
+    if args.fuses is None and args.flash is None and not args.erase:
+        print("No action (erase, flash or fuses)")
+        sys.exit(0)
+
+    if args.verbose:
+        logging.basicConfig(format="%(levelname)s:%(name)s %(message)s",
+                            level=logging.INFO)
+    else:
+        logging.basicConfig(format="%(levelname)s:%(name)s %(message)s",
+                            level=logging.WARNING)
+
+    nvm = UpdiNvmProgrammer(comport=args.comport,
+                            baud=args.baudrate,
+                            device=Device(args.device))
 
-    nvm = UpdiNvmProgrammer(comport=comport, baud=115200, device=device)
-
-    # Retrieve data to write
-    data, start_address = nvm.load_ihex(filename)
-
-    # Enter programming mode
     try:
         nvm.enter_progmode()
-    except Exception:
-        print("Device is locked.  Performing unlock with chip erase.")
+    except:
+        print("Device is locked. Performing unlock with chip erase.")
         nvm.unlock_device()
 
-    # Read and display device info
     nvm.get_device_info()
+    if args.erase:
+        nvm.chip_erase()
+    if args.fuses is not None:
+        for fslist in args.fuses:
+            for fsarg in fslist:
+                if not re.match("^[0-9]+:0x[0-9a-fA-F]+$", fsarg):
+                    print("Bad fuses format {}. Expected fuse_nr:0xvalue".format(fsarg))
+                    continue
+                lst = fsarg.split(":0x")
+                fusenum = int(lst[0])
+                value = int(lst[1], 16)
+                if not _set_fuse(nvm, fusenum, value):
+                    sys.exit(1)
+    if args.flash is not None:
+        _flash_file(nvm, args.flash)
+
+def _flash_file(nvm, filename):
+    data, start_address = nvm.load_ihex(filename)
 
-    # Erase the device
-    nvm.chip_erase()
+    fail=False
 
-    # Program from intel hex file
+    nvm.chip_erase()
     nvm.write_flash(start_address, data)
 
     # Read out again
-    readback = nvm.read_flash(device.flash_start, len(data))
+    readback = nvm.read_flash(nvm.device.flash_start, len(data))
     for i, _ in enumerate(data):
         if data[i] != readback[i]:
-            print("Verify error at location 0x{0:04X}: expected 0x{1:02X} read 0x{2:02X} ".format(i, data[i],
-                                                                                                  readback[i]))
+            print("Verify error at location 0x{0:04X}: expected 0x{1:02X} read 0x{2:02X} ".format(i, data[i], readback[i]))
+            fail=True
+
+    if not fail:
+        print("Programming successful")
+
 
-    # Exit programming mode
-    nvm.leave_progmode()
```

Comment:
Removing this leaves the MCU in progmode when done.
Programming still works, but the code will not run unless power is cycled or a reset is applied.
Adding this in again at the end of main (around line:109) will fix this nicely without breaking anything else.